pyomo/contrib/mindtpy/cut_generation.py

```diff
@@ -93,6 +93,50 @@ def add_oa_cuts(target_model, dual_values, solve_data, config,
                           >= constr.lower)
                 )
 
+def add_ecp_cuts(target_model, solve_data, config,
+                linearize_active=True,
+                linearize_violated=True,
+                linearize_inactive=False):
+    """Linearizes nonlinear constraints.
+
+    For nonconvex problems, turn on 'config.add_slack'. Slack variables will
+    always be used for nonlinear equality constraints.
+    """
+    for constr in target_model.MindtPy_utils.constraint_list:
+        if constr.body.polynomial_degree() in (0, 1):
+            continue
+
+        constr_vars = list(identify_variables(constr.body))
+        jacs = solve_data.jacobians
+
```

Comment:
This was extremely clever. I checked and the code above uses the duals only to determine which direction of the cut should be added if we have an equality constraint. The more I think about it, the more it makes sense to use the fact of `constr.uslack() < 0` and `constr.lslack() < 0` to determine which side of the cut should be added. What do you think @ZedongPeng ?